elements/dash-tree/dash-tree.js

```diff
@@ -0,0 +1,436 @@
+   */
+  class DashTreeGroup extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    connectedCallback() {
+      this.setAttribute('role', 'group');
+    }
+  }
+
+  /**
+   * Define a custom element, `<dash-tree-group>`, and associate it with the
+   * `DashTreeGroup` class.
+   */
+  window.customElements.define('dash-tree-group', DashTreeGroup);
+
+  /**
+   * `DashTree` is responsible for handling user input and updating the
+   * expanded/collapsed and selected state for its children. It also manages
+   * the currently active child using `aria-activedescendant`.
+   */
+  class DashTree extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    connectedCallback() {
+      // The element needs to do some manual input event handling to allow
+      // switching with arrow keys and Home/End.
+      this.addEventListener('keydown', this._onKeyDown);
+      this.addEventListener('click', this._onClick);
+
+      this.setAttribute('role', 'tree');
+      this.setAttribute('tabindex', 0);
+
+      // Before the elements starts booting, it waits for both
+      // the `<dash-tree-item>` and `<dash-tree-group>` to load.
+      Promise.all([
+        customElements.whenDefined('dash-tree-item'),
+        customElements.whenDefined('dash-tree-group'),
+      ]).then(_ => {
+        // Acquire all `DashTreeItem` instances inside the element
+        const treeItems = this._allTreeItems();
+        // If there are no treeItems, then the tree is empty. Abort.
+        if (treeItems.length === 0) return;
+
+        // The element checks if any child has been marked as selected.
+        // If so, it will mark it as the current `aria-activedescendant`
+        // and mark it with a `.selected` class.
+        const selectedTreeItem =
+          treeItems.find(treeItem =>
+            treeItem.getAttribute('aria-selected') === 'true');
+
+        if (selectedTreeItem) {
+          this._focusTreeItem(selectedTreeItem);
+          this._selectTreeItem();
+        }
+      });
+    }
+
+    /**
+     * Returns a list of visible `DashTreeItem` elements. This is useful when
+     * the user wants to try to move to the next or previous item in the list.
+     * If an item is a child of a parent who is set to `aria-expanded=false`
+     * then it is considered invisible and is not added to the list.
+     */
+    _allTreeItems() {
+      const treeItems = [];
+      // A recursive function that visits every child and builds a list
+      // This produces similar results to calling querySelectorAll,
+      // but allows for filtering of the children based on whether or not
+      // their parent is currently expanded.
+      function findTreeItems(node) {
+        for (let el of node.children) {
+          // Ignore children like `<label>`, and `<dash-tree-group>`.
+          // If the element has children, descend into them looking for
+          // more treeitems.
+          if (!isTreeItem(el) && el.firstElementChild) {
```

Comment:
yeah I was able to go with your approach. Just needed to swap out the `return` with a `continue`. Otherwise it _seems_ to work as expected 👍